Avoid Zend dynamic properties prior PHP 8.2
--- a/blitz-0.10.6/blitz.c
+++ b/blitz-0.10.6/blitz.c
@@ -5992,7 +5992,9 @@
     le_blitz = zend_register_list_destructors_ex(blitz_resource_dtor, NULL, "Blitz template", module_number);
 
     INIT_CLASS_ENTRY(blitz_class_entry, "blitz", blitz_functions);
-    blitz_class_entry.ce_flags |= ZEND_ACC_ALLOW_DYNAMIC_PROPERTIES;
+    #if PHP_VERSION_ID >= 80200
+        blitz_class_entry.ce_flags |= ZEND_ACC_ALLOW_DYNAMIC_PROPERTIES;
+    #endif
     zend_register_internal_class(&blitz_class_entry);
     blitz_register_constants(INIT_FUNC_ARGS_PASSTHRU);
 
